src/agent/onefuzz/src/monitor.rs

```diff
@@ -1,33 +1,39 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-use std::{path::PathBuf, pin::Pin, sync::mpsc, time::Duration};
+use std::path::PathBuf;
+use std::sync::{
+    self,
+    mpsc::{Receiver as SyncReceiver, RecvError},
+};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
 
 use anyhow::Result;
-use futures::{
-    stream::{FusedStream, Stream},
-    task::{self, Poll},
-};
 use notify::{DebouncedEvent, Watcher};
-use std::sync::mpsc::TryRecvError;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver};
 
 pub struct DirectoryMonitor {
     dir: PathBuf,
-    rx: mpsc::Receiver<DebouncedEvent>,
+    notify_events: UnboundedReceiver<DebouncedEvent>,
     watcher: notify::RecommendedWatcher,
     terminated: bool,
 }
 
 impl DirectoryMonitor {
     pub fn new(dir: impl Into<PathBuf>) -> Self {
         let dir = dir.into();
-        let (tx, rx) = mpsc::channel();
+        let (notify_sender, notify_receiver) = sync::mpsc::channel();
         let delay = Duration::from_millis(100);
-        let watcher = notify::watcher(tx, delay).unwrap();
+        let watcher = notify::watcher(notify_sender, delay).unwrap();
+
+        // We can drop the thread handle, and it will continue to run until it
+        // errors or we drop the async receiver.
+        let (notify_events, _handle) = into_async(notify_receiver);
```

Comment:
do we need this indirection ? Why not just read from the the `notify_sender` ?

Reply:
`notify_receiver` is not async, so to block on receiving messages (and thus receive OS filesystem notification pushes), we need to do it off-thread. The `notify::Watcher` requires a `std::sync::mpsc` channel, so we can't just pass an async `Sender` to the `Watcher`.

Reply:
got it.
In that case i think we should use [try_rcv](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.try_recv) in the implementation of `into_async` to allow us to yield the task when there is nothing instead of blocking

Reply:
Actually I think we can even eliminated the spawned task and channel by using a logic similar to the code that you deleted in `src/agent/onefuzz-agent/src/local/common.rs` .

Reply:
We are still yielding, because we only block in a worker thread.

This lets us (1) use the FS notify-push blocking API that we get from the `notify` crate in a way that doesn't block our async scheduling, but (2) use an async-aware `UnboundedReceiver` in the impl of `DirectoryMonitor::next_file()` (which is now an `async` method). So our API _does_ yield to the async runtime if there are no new messages, as desired.

Reply:
As per our discussions. The goal with my proposed change is remove the allocation of an OS thread which could potentially be blocking. 
My assumption is that by using the tokio task we consume much less resource that allocating the OS thread.

Reply:
Agreed. I'll revisit this again to try and bridge the sync/async receivers with less overhead.